core/controller/src/main/scala/org/apache/openwhisk/core/controller/Actions.scala

```diff
@@ -333,12 +372,74 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
    * - 500 Internal Server Error
    */
   override def remove(user: Identity, entityName: FullyQualifiedEntityName)(implicit transid: TransactionId) = {
-    deleteEntity(WhiskAction, entityStore, entityName.toDocId, (a: WhiskAction) => Future.successful({}))
+    parameter('version.as[SemVer] ?) { version =>
+      onComplete(WhiskActionVersionList.get(entityName, entityStore)) {
+        case Success(results) =>
+          version match {
+            case Some(_) =>
+              val docId = results.matchedDocId(version).getOrElse(entityName.toDocId)
+              deleteEntity(
+                WhiskAction,
+                entityStore,
+                docId,
+                (a: WhiskAction) => Future.successful({}),
+                postProcess = Some { action: WhiskAction =>
+                  WhiskActionVersionList.deleteCache(entityName)
+                  complete(OK, action)
+                })
+            case None =>
+              val fs =
+                if (results.versions.isEmpty)
+                  Seq(WhiskAction.get(entityStore, entityName.toDocId) flatMap { entity =>
+                    WhiskAction.del(entityStore, entity.docinfo).map(_ => entity)
+                  })
+                else
+                  results.versions.values
```

Comment:
I think this is a little scary to delete all versions if a version accidentally isn't supplied. Should there be a delete all parameter flag otherwise return an error?

Also is it possible that you have a list of versions and then from backwards compatibility still have an action with no version? Does that result in the action with no version to get left over after the delete?

Reply:
ok

since it will fetch a version-id mappings at first, so even for action with no version, we can get the id of it and delete action by id